ToolBox/TradingEconomicsDataDownloader/TradingEconomicsIndicatorDownloader.cs

```diff
@@ -67,44 +94,94 @@ public override bool Run()
                     try
                     {
                         var endUtc = startUtc.AddMonths(1).AddDays(-1);
+
+                        if (availableFiles.Contains(endUtc))
+                        {
+                            Log.Trace($"TradingEconomicsIndicatorDownloader.Run(): Skipping data because it already exists for month: {startUtc:MMMM}");
+                            startUtc = startUtc.AddMonths(1);
+                            continue;
+                        }
+
+                        Log.Trace($"TradingEconomicsIndicatorDownloader.Run(): Collecting data for indicator: {indicator} - from {startUtc:yyyy-MM-dd} to {endUtc:yyyy-MM-dd}");
+
+                        _requestGate.WaitToProceed(TimeSpan.FromSeconds(1));
+
                         var content = Get(startUtc, endUtc).Result;
                         var collection = JsonConvert.DeserializeObject<List<TradingEconomicsIndicator>>(content);
 
                         data.AddRange(collection);
-
                         startUtc = startUtc.AddMonths(1);
                     }
                     catch (Exception e)
                     {
-                        Log.Error(e, $"TradingEconomicsIndicatorsDownloader(): Error parsing data for date {startUtc:yyyyMMdd}");
+                        Log.Error(e, $"TradingEconomicsIndicatorDownloader.Run(): Error parsing data for date {startUtc:yyyyMMdd}");
                         return false;
                     }
                 }
 
-                Log.Trace($"TradingEconomicsIndicatorsDownloader(): {data.Count} {indicator} indicator entries read in {stopwatch.Elapsed}");
+                Log.Trace($"TradingEconomicsIndicatorDownloader.Run(): {data.Count} {indicator} indicator entries read in {stopwatch.Elapsed}");
 
-                foreach (var kvp in data.GroupBy(GetFileName))
-                {
-                    var path = Path.Combine(_destinationFolder, kvp.Key);
-                    var zipPath = path.Replace(".json", ".zip");
+                // Return status code. We default to `true` so that we can identify if an error occured during the loop
+                var status = true;
 
-                    try
+                Parallel.ForEach(data.GroupBy(GetTicker),
```

Comment:
Not sure if possible, but there seems to be some logical code duplication in this method and `GetTicker` between the different `TradingEconomicsYYYYYYDownloaders` maybe there could be a base downloader with some of this logic?